### ISSUE TITLE
Using uninterpreted sorts with SyGuS

### ISSUE THREAD

**Issue by murali24adithya** [CVC4/CVC4#5379]
I am trying to use CVC4Sy to synthesise an expression against a set of defined functions. The mathematically correct definition of the functions uses a finite uninterpreted domain which I have modeled using an uninterpreted sort in the attached file with-sort.txt. This does not give me any results for a long time when I just run cvc4 --lang=sygus2 <filename>. However, when I replace all the constants of the finite domain with natural numbers (see with-int.txt) I get results immediately, again without using any special options.

How can I get the same performance with uninterpreted sorts as well? Are there any special options or is it the case that in general uninterpreted sorts are not a good modeling choice to give to cvc4sy?

Minimal example:
[with-int.txt](https://github.com/CVC4/CVC4/files/5478309/with-int.txt)
[with-sort.txt](https://github.com/CVC4/CVC4/files/5478310/with-sort.txt)

**Comment by ajreynol** [comment]
Thanks for the benchmark.

In general, uninterpreted sorts do not mix well with SyGuS in cvc4.   

The high-level reason is that SyGuS is based on accumulating counterexamples to the given conjecture in the form of models for quantifier free queries.  When uninterpreted sorts are present, this becomes more complicated, since the model actually is not just a mapping from variables to value, but also an interpretation of uninterpreted sorts to (finite) sets.

In summary, I would generally recommend avoiding uninterpreted sorts + SyGuS.  Regardless, we should try to improve the robustness on inputs with uninterpreted sorts (e.g. provide warnings/errors, or otherwise come up with a robust solution to handling them).

**Comment by murali24adithya** [comment]
Thank you so much!
